activesupport/lib/active_support/core_ext/object/json.rb

```diff
@@ -239,9 +238,20 @@ def as_json(options = nil)
   end
 end
 
-class IPAddr # :nodoc:
-  def as_json(options = nil)
-    to_s
+unless defined?(IPAddr::VERSION) && Gem::Version.new(IPAddr::VERSION) >= Gem::Version.new("1.2.7")
+  # If use ipaddr gem v1.2.7+, use native `IPAddr#as_json`
+  require "ipaddr"
```

Comment:
Would you tell me the reason why `require "ipaddr"` here? Without `require "ipaddr"` the line 241 looks always returns true regardless of IPAddr version is 1.2.7 or lower.

- Ruby 3.3.5 with ipaddr 1.2.6
```ruby
$ ruby -v
ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [x86_64-linux]
$ gem info ipaddr

*** LOCAL GEMS ***

ipaddr (1.2.6)
    Authors: Akinori MUSHA, Hajimu UMEMOTO
    Homepage: https://github.com/ruby/ipaddr
    Licenses: Ruby, BSD-2-Clause
    Installed at (default): /home/yahonda/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0

    A class to manipulate an IP address in ruby
$ irb -w
irb(main):001> defined?(IPAddr::VERSION)
=> nil
irb(main):002> Gem::Version.new(IPAddr::VERSION)
(irb):2:in `<main>': uninitialized constant IPAddr (NameError)

Gem::Version.new(IPAddr::VERSION)
                 ^^^^^^
	from <internal:kernel>:187:in `loop'
	from /home/yahonda/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/irb-1.14.1/exe/irb:9:in `<top (required)>'
	from /home/yahonda/.rbenv/versions/3.3.5/bin/irb:25:in `load'
	from /home/yahonda/.rbenv/versions/3.3.5/bin/irb:25:in `<main>'
irb(main):003>
irb(main):004* unless defined?(IPAddr::VERSION) && Gem::Version.new(IPAddr::VERSION) >= Gem::Version.new("1.2.7")
irb(main):005*   p 'foo'
irb(main):006> end
"foo"
=> "foo"
irb(main):007>
```

- Ruby 3.4.0dev with ipaddr 1.2.7
```ruby
$ ruby -v
ruby 3.4.0dev (2024-10-19T13:00:02Z master 1c653cabfc) +PRISM [x86_64-linux]
$ gem info ipaddr

*** LOCAL GEMS ***

ipaddr (1.2.7)
    Authors: Akinori MUSHA, Hajimu UMEMOTO
    Homepage: https://github.com/ruby/ipaddr
    Licenses: Ruby, BSD-2-Clause
    Installed at (default): /home/yahonda/.rbenv/versions/trunk/lib/ruby/gems/3.4.0+0

    A class to manipulate an IP address in ruby
$ irb -w
... snip warnings...
irb(main):001> defined?(IPAddr::VERSION)
=> nil
irb(main):002> Gem::Version.new(IPAddr::VERSION)
(irb):2:in '<main>': uninitialized constant IPAddr (NameError)
	from <internal:kernel>:168:in 'Kernel#loop'
	from /home/yahonda/.rbenv/versions/trunk/lib/ruby/gems/3.4.0+0/gems/irb-1.14.1/exe/irb:9:in '<top (required)>'
	from /home/yahonda/.rbenv/versions/trunk/bin/irb:25:in 'Kernel#load'
	from /home/yahonda/.rbenv/versions/trunk/bin/irb:25:in '<main>'
irb(main):003>
irb(main):004* unless defined?(IPAddr::VERSION) && Gem::Version.new(IPAddr::VERSION) >= Gem::Version.new("1.2.7")
irb(main):005*   p 'foo'
irb(main):006> end
"foo"
=> "foo"
irb(main):007>
```

Reply:
How about restoring `require "ipaddr"` at the line 6 and using `unless IPAddr.method_defined?(:as_json)` because `as_json` method added in IPAddr 1.2.7 via https://github.com/ruby/ipaddr/pull/77 ?

Reply:
@yahonda Thanks to review.
> How about restoring require "ipaddr" at the line 6
It is mistake. I have fixed it.

If use `unless IPAddr.method_defined?(:as_json)`, test failed.
Because `IPAddr.method_defined?(:as_json)` return true, it seems to defined `as_json` of `Module`/`Object` class in this file.
https://github.com/rails/rails/blob/7c02b81ec3025d1067ef0065e377d27db0eba98f/activesupport/lib/active_support/core_ext/object/json.rb#L52-L66

Reply:
> Because IPAddr.method_defined?(:as_json) return true, it seems to defined as_json of Module/Object class in this file.

How about setting inherit argument as false to see if the as_json method is defined in the IPAddr, not to include inherited or included ones?

```ruby
unless IPAddr.method_defined?(:as_json, inherit=false)
```

https://docs.ruby-lang.org/en/3.1/Module.html#method-i-method_defined-3F

Reply:
Thanks to comments. It have fixed to use `unless IPAddr.method_defined?(:as_json, inherit=false)`.